modules/core/src/scenegraph/gltf-instantiator/gltf-instantiator.js

```diff
@@ -27,75 +27,95 @@ export default class GLTFInstantiator {
 
   createNode(gltfNode) {
     const gltfMeshes = gltfNode.children || [];
-    const meshes = gltfMeshes.map(node => this.createMesh(node));
-    // Option: Avoid group if single Mesh
-    let node = (this.options.removeEmptyGroups && meshes.length === 1) && meshes[0];
-    node = node || new Group({
+    const children = gltfMeshes.map(child => this.createNode(child));
+    // TODO: Check if we can have both children nodes and meshes!
+    if (gltfNode.mesh) {
+      children.push(this.createMesh(gltfNode.mesh));
+    }
+    const node = new Group({
       id: gltfNode.name || gltfNode.id,
-      children: meshes
+      children
     });
+    if (gltfNode.matrix) {
+      node.setMatrix(gltfNode.matrix);
+    }
     return node;
   }
 
   createMesh(gltfMesh) {
+    // TODO: avoid changing the gltf
     if (!gltfMesh._mesh) {
       const gltfPrimitives = gltfMesh.primitives || [];
       const primitives = gltfPrimitives.map(
         (gltfPrimitive, i) => this.createPrimitive(gltfPrimitive, i, gltfMesh)
       );
-      // Option: avoid group if single Primitive
-      let mesh = (this.options.removeEmptyGroups && primitives.length === 1) && primitives[0];
-      mesh = mesh || new Group({
+      const mesh = new Group({
         id: gltfMesh.name || gltfMesh.id,
         children: primitives
       });
       gltfMesh._mesh = mesh;
     }
+
     return gltfMesh._mesh;
   }
 
   createPrimitive(gltfPrimitive, i, gltfMesh) {
     const attributes = this.createAttributes(gltfPrimitive.attributes, gltfPrimitive.indices);
     // TODO - handle gltfPrimitive.material
-    return new Model(this.gl, {
+    const model = new Model(this.gl, {
       id: gltfPrimitive.name || `${gltfMesh.name || gltfMesh.id}=${i}`,
-      attributes,
-      drawMode: gltfPrimitive.mode
+      drawMode: gltfPrimitive.mode || 4,
+      vertexCount: 2046 // from indices count
     });
+    model.setProps({attributes});
+    this.options.getImage(0).then(img => model.setUniforms({ tex: img }));
```

Comment:
I think it would be nice to have a single function that resolves all promises in a loaded glTF, rather than having to wait for them one by one like this.

That way a loaded file could be processed without dealing with promises. Of course if we want to support incremental loading, we'd need to do it like this.

The promise resolving function could be extended to handle URI loading etc.

Reply:
But then I have to change the loader.